Add rust survey post

nellshamrell opened the pull request:
This is a better formatted and more refined draft of a post about the 2019 survey. I would LOVE to get some help for the sections that are marked with "ADD COPY".

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

XAMPPRocky commented:
@nellshamrell It seems like there's two files?

nellshamrell commented:
Whoops! Accidentally left the old draft in. It is fixed now :)

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+## Who is using Rust and what for?
+
+![How long have you worked in Rust](/images/2020-03-RustSurvey/7-How-Long-Worked-With-Rust.png)
```
I think if we do include pie charts, I would prefer that we have labels on each sector for the %age, given the research I've seen that suggests interpreting them correctly is otherwise quite hard :)

I would also like to see if we can perhaps avoid the close-to-duplicate colors here? There's ~3 blues that are quite similar looking at a glance, though perhaps this is related to the resolution issue I noted elsehwere.

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+![Size of summed Rust projects](/images/2020-03-RustSurvey/10-Size-Of-Summed-projects.png)
+
+The majority of Rust projects (43%) are 1,000-10,000 lines of code. Rust projects of medium to large size (those totaling over 10k and 100k lines of code respectively) continue to trend higher. They have grown from 8.9% in 2016, to 16% in 2017, to 23% in 2018, to **34%** in 2019.
```
I'm not quite sure what this sentence is meant to mean -- perhaps just the following? I think presumably we're in just the "all projects greater than 10k" range.

```suggestion
The majority of Rust projects (43%) are 1,000-10,000 lines of code. Rust projects of medium to large size (those totaling over 10k lines of code) continue to trend higher. They have grown from 8.9% in 2016, to 16% in 2017, to 23% in 2018, to **34%** in 2019.
```

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+![How often do you use Rust](/images/2020-03-RustSurvey/11-How-Often-Use-Rust.png)
+
+Rust daily usage has trended slightly upward at 27.63% (it was just under 25% last year and 17.5% on 2017). Users who use Rust daily or weekly has also continued to rend slight upward. This year it was 68.52%, last year it was 66.4%, and in 2017 it was 60.8%.
```
```suggestion
Rust daily usage has trended slightly upward at 27.63% (it was just under 25% last year and 17.5% on 2017). Daily or weekly usage has also continued to trend slight upward. This year it was 68.52%, last year it was 66.4%, and in 2017 it was 60.8%.
```

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+![How would you rate your Rust expertise](/images/2020-03-RustSurvey/12-How-Rate-Rust-Expertise.png)
+
+We also asked users how they would rate their Rust Expertise - there is a clear peak around "7".
```
This point seems to be missing the elaboration on movement over prior years that we typically have; looking back here is the graph from 2018 and this years (respectively); it looks like this has on average not changed though we've had slightly less responses? I think noting the lack of change would be good, IMO that's a little concerning -- though as I think @nikomatsakis mentioned elsewhere, it would be nice to correlate this with "length of time using Rust" and "frequency of use". Hopefully, people are trending towards expert usage, especially if they use it weekly/daily.



![image](https://user-images.githubusercontent.com/5047365/78194576-8fabed80-744b-11ea-94f4-1970d1a33a35.png)

![image](https://user-images.githubusercontent.com/5047365/78194580-933f7480-744b-11ea-92cf-984d925457ed.png)

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+For users who had used Rust before but had stopped using it, most had stopped using it within a year of starting.
+
+![Why did you stop using Rust?](images/2020-03-RustSurvey/4-Why-Stopped-Rust.png)
```
```suggestion
![Why did you stop using Rust?](/images/2020-03-RustSurvey/4-Why-Stopped-Rust.png)
```

Mark-Simulacrum replied:
I'm not sure what we can do about this, but the text on the bottom for these graphs runs together making it hard to read.

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+When we asked why someone had stopped using Rust, the most common response was "My company doesn't use Rust" - which indicates Rust adoption is still the biggest reason. After that, learning curve, lack of needed libraries, being slowed down by switching to Rust, and lack of IDE support were the most common reasons a user stopped using Rust.
+
+![Why have you never used Rust?](images/2020-03-RustSurvey/5-Why-Never-Used-Rust.png)
```
```suggestion
![Why have you never used Rust?](/images/2020-03-RustSurvey/5-Why-Never-Used-Rust.png)
```

Mark-Simulacrum replied:
Hm, why did we split out "haven't but want to" to the right?

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+For users who indicated they had never used Rust before, most indicated either "I haven't learned Rust yet, but I want to" or "My company doesn't use Rust" - again pointing to adoption as the main hurdle.
+
+![I would use Rust more often if...](images/2020-03-RustSurvey/20-Would-use-rust-more-often-if.png)
```
This graph is tiny compared to the other bar charts in this section, though I'm not sure why. It also trails off on many of the labels into `...`.

Mark-Simulacrum replied:
```suggestion
![I would use Rust more often if...](/images/2020-03-RustSurvey/20-Would-use-rust-more-often-if.png)
```

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+We also asked users what would lead them to use Rust more often. Most indicated they would use Rust more if their company adopted it, if Rust had more libraries that they need, and if IDE support was better. The most common reasons after those pointed to a need to improve the learning curve and interoperability. 
+
+As adoption seemed to be the biggest problem preventing some respondents from using Rust, let's dive deeper into it. 
```
One thought -- it might make sense to change the absolute bars to percentages of total respondents to that question, or perhaps the survey in general (not sure what makes more sense).

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+First, we asked what would make Rust more appealing to your company and convince them to adopt it.
+
+![How can we make Rust more appealing to your company](images/2020-03-RustSurvey/21-Rust-more-appealing.png)
```
Title on this graph is all uppercase for some reason.

Mark-Simulacrum replied:
We should be careful to indicate what the percentages on the y-axis here mean (i.e., of what).

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+* serde
+* rand
+* tokio
+* async
```
Perhaps some sort of filtering gone wrong and meant to be async-std?

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+* 5.2% indicated "None"
+* 21.4% indicated "I don't know"
+
+[Add COPY/analysis here]
```
note: in particular, we should try to tie the libraries above -- I think all of which, with the exception of serde/clap/regex/lazy_static, are very low level -- you would generally want to wrap them in something else to get the same nicety of a framework-like experience in another language.

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+![Which Rust version do you use for your applications?](images/2020-03-RustSurvey/16-Which-Rust-Version.png)
+
+The vast majority of users use the current stable version of Rust (63%). Surprisingly, the number of users using the Nightly compiler in their workflow is down at 20%. Last year it was at over 56%.
```
I think noting the exact previous percentages for stable and nightly and the new percentages would be quite nice. I would also want to be careful -- what are these percentages of? Total survey respondents? Respondents to this question?

nellshamrell replied:
Added in the exact percentages for this year, and indicated that these percentages are of all respondents (though, as noted now, respondents could select more than one version)

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+Rust is well known for its significant learning curve. 
+
+![How would you rate your Rust expertise](/images/2020-03-RustSurvey/13-Rust-Expertise-and-Time-In-Rust.png)
```
It loosely seems like the width of the bars in this graph is changing as we go to the right? Maybe it's my eyes...

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+Over the years some users have expressed a desire for Rust to be more interoperable with other languages.
+
+![What languages would you want to use with Rust](/images/2020-03-RustSurvey/23-interoperability-languages.png)
```
Was this an exclusive question? I seem to recall being able to select more than one -- in that case a pie chart would be the wrong choice I think? Not sure exactly what the percentages mean (i.e. they're of some sort of "total number of selected choices" not "people selecting")

Mark-Simulacrum commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+![What languages would you want to use with Rust](/images/2020-03-RustSurvey/23-interoperability-languages.png)
+
+When we asked users what languages they would want to be interoperable with Rust, there was a wide spread of answers, but C dominates.
```
I think noting the (to me *very* surprising) result that R is dominant behind C - with what looks like a 5% gap to C++ - is good.

XAMPPRocky commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+The common blockers that people mention to participating is that they have social anxiety, and accessibility. One of the common reasons mentioned was that some resources are hard to read for people with dyslexia.
+
+These issues are definitely ones we want to address in 2020 and beyond.
```
Reposting on this PR. cc @nellshamrell @nikomatsakis I've made a document containing three lists of selected quotes for the question of "What actions could we take to make you feel more welcome?". Are there any specific quotes that particularly stand out to either of you that we should include?

https://paper.dropbox.com/doc/What-actions-could-we-take-to-make-you-feel-more-welcome--AxQVDWInzOSddn~2_y5hqYrCAg-vsqRYS9QYiOw0fKh1I2qK

nellshamrell replied:
Thanks so much! I've pulled several quotes from there into the draft.

XAMPPRocky commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+## Who is using Rust and what for?
+
+![How long have you worked in Rust](/images/2020-03-RustSurvey/7-How-Long-Worked-With-Rust.png)
```
This is a problem with Google Sheets, when you download a PNG it's at 1x and you can't download a high res version. What we could do though is replace the PNGs with SVGs. @nellshamrell If you send me access to the sheet that contains these charts, I can go through and update charts based on marks feedback?

nellshamrell replied:
I pulled these straight from a dropbox paper (they were already generated) - I'm unsure if I have access to the Google sheet they were generated from, but I will check.

wezm commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+Greetings Rustaceans!
+
+We are happy to present the results of our fourth annual survey of our Rust community. Before we did into the analysis, we want to give a big "thank you!" to all of the people who took the time to respond. You are vital to Rust continuing to improve year after year!
```
```suggestion
We are happy to present the results of our fourth annual survey of our Rust community. Before we dig into the analysis, we want to give a big "thank you!" to all of the people who took the time to respond. You are vital to Rust continuing to improve year after year!
```

nikomatsakis commented on posts/2020-03-31-Rust-survey-2019.md:
```diff
@@ -0,0 +1,216 @@
+
+![How long have you been unproductive](/images/2020-03-RustSurvey/9-How-Long-Unproductive.png)
+
+Additionally, many users indicated that they were not feeling comfortable in Rust - even after over 3 months of use. This is likely a result of Rust's high learning curve.
```
Nit: I would delete the last sentence and merge this with the next paragraph, I think.

nellshamrell commented:
Update - thank you for the feedback! I believe I have addressed all the comments except for the ones re: graphs - I am working with @XAMPPRocky to get the graphs updated and refined :)

nikomatsakis commented:
Sorry if that feedback is no longer relevant.

nellshamrell commented:
Alright - I've addressed the feedback as best as a can.  Huge, HUGE thanks to @XAMPPRocky for generating and re-generating the graphs!

nikomatsakis commented:
I just want to say to both @nellshamrell and @XAMPPRocky that this is really well done. 🎉
